File: webscrambles/src/main/java/net/gnehzr/tnoodle/server/webscrambles/ScrambleHandler.java
package net.gnehzr.tnoodle.server.webscrambles;

import com.itextpdf.text.DocumentException;
import net.gnehzr.tnoodle.server.SafeHttpServlet;
import net.lingala.zip4j.exception.ZipException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Date;
import java.util.Map;

import static net.gnehzr.tnoodle.utils.GsonUtils.GSON;

@SuppressWarnings("serial")
public class ScrambleHandler extends SafeHttpServlet {

    public ScrambleHandler() {}

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
        String[] path = parsePath(request.getPathInfo());

        if(path.length == 0) {
            response.sendError(404);
        } else {
            Date generationDate = new Date();

            Map<String, String> query = parseQuery(request.getQueryString());

            // TODO - this means you can't have a round named "seed" or "showIndices" or "callback" or "generationUrl"!
            String seed = query.remove("seed");
            String generationUrl = query.remove("generationUrl");
            boolean showIndices = query.remove("showIndices") != null;
            query.remove("callback");

            String globalTitle, ext;
            int lastDot = path[0].lastIndexOf(".");
            if(lastDot < 0) {
                throw new ServletException(new InvalidScrambleRequestException("No extension specified"));
            }
            globalTitle = path[0].substring(0, lastDot);
            ext = path[0].substring(lastDot+1);

            try {
                if(ext == null || ext.equals("txt")) {
                    // Note that we parse the scramble requests *after* checking the extension.
                    // This way, someone who makes a request for "/scramble/foo.bar" will get a warning about
                    // the ".bar" extension, rather than incorrect scramble requests.
                    ScrambleRequest[] scrambleRequests = ScrambleRequest.parseScrambleRequests(query, seed);

                    StringBuilder sb = new StringBuilder();
                    for(ScrambleRequest scrambleRequest : scrambleRequests) {
                        for(int j = 0; j < scrambleRequest.copies; j++) {
                            for(int i = 0; i < scrambleRequest.scrambles.length; i++) {
                                String scramble = scrambleRequest.scrambles[i];
                                if(showIndices) {
                                    sb.append((i+1)).append(". ");
                                }
                                // We replace newlines with spaces
                                sb.append(scramble.replaceAll("\n", " "));
                                sb.append("\r\n");
                            }
                        }
                    }
                    sendText(request, response, sb.toString());
                } else if(ext.equals("json")) {
                    ScrambleRequest[] scrambleRequests = ScrambleRequest.parseScrambleRequests(query, seed);
                    sendJSON(request, response, GSON.toJson(scrambleRequests));
                } else if(ext.equals("pdf")) {
                    ScrambleRequest[] scrambleRequests = ScrambleRequest.parseScrambleRequests(query, seed);
                    ByteArrayOutputStream totalPdfOutput = ScrambleRequest.requestsToPdf(globalTitle, generationDate, scrambleRequests, null);
                    response.setHeader("Content-Disposition", "inline");

                    // Workaround for Chrome bug with saving PDFs:
                    //  https://bugs.chromium.org/p/chromium/issues/detail?id=69677#c35
                    response.setHeader("Cache-Control", "public");

                    sendBytes(request, response, totalPdfOutput, "application/pdf");
                } else if(ext.equals("zip")) {
                    ScrambleRequest[] scrambleRequests = ScrambleRequest.parseScrambleRequests(query, seed);

                    ByteArrayOutputStream baosZip = ScrambleRequest
                            .requestsToZip(getServletContext(), globalTitle, generationDate, scrambleRequests, seed, generationUrl, null);
                    sendBytes(request, response, baosZip, "application/zip");
                } else {
                    throw new InvalidScrambleRequestException("Invalid extension: " + ext);
                }
            } catch (InvalidScrambleRequestException | DocumentException | ZipException e) {
                throw new ServletException(e);
            }
        }
    }
}

File: webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/ScrambleHandler.kt
package org.worldcubeassociation.tnoodle.server.webscrambles

import io.ktor.application.call
import io.ktor.http.ContentType
import io.ktor.response.header
import io.ktor.response.respond
import io.ktor.response.respondBytes
import io.ktor.response.respondText
import io.ktor.routing.Routing
import io.ktor.routing.get
import io.ktor.util.toMap
import org.worldcubeassociation.tnoodle.server.RouteHandler

import java.util.Date

object ScrambleHandler : RouteHandler {
    override fun install(router: Routing) {
        router.get("/scramble/{filename}") {
            val filename = call.parameters["filename"]!!

            val generationDate = Date()

            val query = call.request.queryParameters.toMap().mapValues { it.value.first() }.toMutableMap()

            // TODO - this means you can't have a round named "seed" or "showIndices" or "callback" or "generationUrl"!
            val seed = query.remove("seed")
            val generationUrl = query.remove("generationUrl")
            val showIndices = query.remove("showIndices") != null

            query.remove("callback")

            val lastDot = filename.lastIndexOf(".")

            if (lastDot < 0) {
                throw InvalidScrambleRequestException("No extension specified")
            }

            val globalTitle = filename.substring(0, lastDot)

            // Note that we parse the scramble requests *after* checking the extension.
            // This way, someone who makes a request for "/scramble/foo.bar" will get a warning about
            // the ".bar" extension, rather than incorrect scramble requests.
            val scrambleRequests = ScrambleRequest.parseScrambleRequests(query, seed)

            when (val ext = filename.substring(lastDot + 1)) {
                "txt" -> {
                    val sb = StringBuilder()
                    for (scrambleRequest in scrambleRequests) {
                        for (j in 0 until scrambleRequest.copies) {
                            for (i in scrambleRequest.scrambles.indices) {
                                val scramble = scrambleRequest.scrambles[i]

                                if (showIndices) {
                                    sb.append(i + 1).append(". ")
                                }

                                // We replace newlines with spaces
                                sb.append(scramble.replace("\n".toRegex(), " "))
                                sb.append("\r\n")
                            }
                        }
                    }

                    call.respondText(sb.toString())
                }
                "json" -> {
                    call.respond(scrambleRequests)
                }
                "pdf" -> {
                    val totalPdfOutput = ScrambleRequest.requestsToPdf(globalTitle, generationDate, scrambleRequests, null)
                    call.response.header("Content-Disposition", "inline")

                    // Workaround for Chrome bug with saving PDFs:
                    //  https://bugs.chromium.org/p/chromium/issues/detail?id=69677#c35
                    call.response.header("Cache-Control", "public")

                    call.respondBytes(totalPdfOutput.toByteArray(), ContentType.Application.Pdf)
                }
                "zip" -> {
                    val baosZip = ScrambleRequest
                        .requestsToZip(getServletContext(), globalTitle, generationDate, scrambleRequests, seed, generationUrl, null)

                    call.respondBytes(baosZip.toByteArray(), ContentType.Application.Zip)
                }
                else -> throw InvalidScrambleRequestException("Invalid extension: $ext")
            }
        }
    }
}
